File: flutter_photo_organizer/lib/features/chat/domain/chatbot.dart
// Chatbot Service

import 'package:image/image.dart' as img;
import '../models/photo_models.dart';
import '../database/database_service.dart';
import 'image_processor.dart';

class PhotoChatbot {
  final DatabaseService _db;
  final ImageProcessor _processor;
  final List<ChatMessage> _conversationHistory = [];

  PhotoChatbot({
    DatabaseService? database,
    ImageProcessor? processor,
  }) : _db = database ?? DatabaseService.instance,
       _processor = processor ?? ImageProcessorImpl();

  List<ChatMessage> get conversationHistory => List.unmodifiable(_conversationHistory);

  Future<String> chat(String message) async {
    // Add user message to history
    final userMsg = ChatMessage(
      id: DateTime.now().millisecondsSinceEpoch,
      role: 'user',
      content: message,
      timestamp: DateTime.now().toIso8601String(),
    );
    _conversationHistory.add(userMsg);
    await _db.storeChatMessage(userMsg);

    // Generate response
    final response = await _generateResponse(message);

    // Add assistant message to history
    final assistantMsg = ChatMessage(
      id: DateTime.now().millisecondsSinceEpoch + 1,
      role: 'assistant',
      content: response,
      timestamp: DateTime.now().toIso8601String(),
    );
    _conversationHistory.add(assistantMsg);
    await _db.storeChatMessage(assistantMsg);

    return response;
  }

  Future<String> askAboutPhoto(String question, ImageAnalysis analysis) async {
    final context = _buildPhotoContext(analysis);
    final prompt = 'Photo analysis: $context\n\nQuestion: $question';
    return await _generateResponse(prompt);
  }

  Future<String> _generateResponse(String message) async {
    final msg = message.toLowerCase();
    final now = DateTime.now();
    final totalPhotos = await _db.getTotalPhotoCount();

    // Greeting
    if (msg.startsWith('hello') || msg.startsWith('hi') || msg.startsWith('hey')) {
      return _buildGreeting(totalPhotos, now);
    }

    // Help
    if (msg.contains('help')) {
      return _buildHelp();
    }

    // Stats
    if (msg.contains('how many') || msg.contains('total') || msg.contains('count') || msg.contains('stats')) {
      return await _buildStatsResponse(totalPhotos);
    }

    // Recent photos
    if (msg.contains('recent') || msg.contains('latest') || msg.contains('newest')) {
      return await _buildRecentPhotosResponse();
    }

    // Scenes
    if (msg.contains('scene') || msg.contains('type') || msg.contains('category') || 
        msg.contains('indoor') || msg.contains('outdoor') || msg.contains('beach') || 
        msg.contains('mountain')) {
      return await _buildScenesResponse();
    }

    // Locations
    if (msg.contains('location') || msg.contains('place') || msg.contains('where') || 
        msg.contains('city') || msg.contains('country')) {
      return await _buildLocationsResponse();
    }

    // People
    if (msg.contains('person') || msg.contains('people') || msg.contains('face') || 
        msg.contains('who') || msg.contains('recognize')) {
      return await _buildPeopleResponse();
    }

    // Date/Time
    if (msg.contains('today') || msg.contains('date') || msg.contains('day') || msg.contains('time')) {
      return 'Today is ${_formatDate(now)} at ${_formatTime(now)}.';
    }

    // Organize
    if (msg.contains('sort') || msg.contains('organize')) {
      return 'I can help organize photos. Use the Organize screen to start sorting your photos by persons, locations, objects, or events.';
    }

    // Thanks
    if (msg.contains('thank') || msg.contains('thanks') || msg.contains('nice')) {
      return "You're welcome! Let me know if you need anything else with your photos.";
    }

    // Default response
    return _buildDefaultResponse(totalPhotos);
  }

  String _buildGreeting(int totalPhotos, DateTime now) {
    return 'Hello! I\'m your photo assistant. It\'s ${_formatDate(now)}. '
        'I can help you explore your $totalPhotos photos. '
        'Try asking about people, places, or specific photos. Type "help" to see what I can do.';
  }

  String _buildHelp() {
    return 'I can help you with:\n'
        '• Find photos by person: "Show me photos with [name]"\n'
        '• Find photos by place: "Show photos from [location]"\n'
        '• Find photos by time: "Show photos from last month"\n'
        '• Photo stats: "How many photos do I have?"\n'
        '• Organize photos: "Organize my photos"\n'
        '• Recent photos: "Show recent photos"\n'
        '• Scene info: "What scenes are in my photos?"';
  }

  Future<String> _buildStatsResponse(int totalPhotos) async {
    final locations = await _db.getAllLocations();
    final faces = await _db.getAllFaces();
    final scenes = await _db.getAllScenes();
    
    final parts = <String>['You have $totalPhotos photos in your library.'];
    if (locations.isNotEmpty) {
      parts.add('They span ${locations.length} different locations.');
    }
    if (faces.isNotEmpty) {
      final known = faces.where((f) => f.name != 'unknown').length;
      parts.add('I\'ve identified $known different people.');
    }
    if (scenes.isNotEmpty) {
      parts.add('There are ${scenes.length} different scene types.');
    }
    return parts.join(' ');
  }

  Future<String> _buildRecentPhotosResponse() async {
    final photos = await _db.searchPhotos('', limit: 5);
    if (photos.isEmpty) {
      return 'No photos found in the database yet.';
    }
    
    final lines = <String>['Here are your most recent photos:'];
    for (final photo in photos) {
      final name = photo.filename;
      final date = photo.dateTaken ?? 'unknown date';
      lines.add('• $name ($date)');
    }
    return lines.join('\n');
  }

  Future<String> _buildScenesResponse() async {
    final scenes = await _db.getAllScenes();
    if (scenes.isEmpty) {
      return 'No scenes analyzed yet. Run the organizer first.';
    }
    
    final lines = <String>['Here are the scene types in your photos:'];
    for (final scene in scenes) {
      final count = await _db.countPhotosByScene(scene['name'] as String);
      lines.add('• ${scene['name']}: $count photos');
    }
    return lines.join('\n');
  }

  Future<String> _buildLocationsResponse() async {
    final locations = await _db.getAllLocations();
    if (locations.isEmpty) {
      return 'No locations found yet. Run the organizer first.';
    }
    
    final lines = <String>['Here are the locations in your photos:'];
    for (final loc in locations) {
      lines.add('• ${loc['name']} (${loc['photo_count']} photos)');
    }
    return lines.join('\n');
  }

  Future<String> _buildPeopleResponse() async {
    final faces = await _db.getAllFaces();
    if (faces.isEmpty) {
      return 'No faces identified yet. Run the organizer first.';
    }
    
    final lines = <String>['Here are the people I\'ve identified:'];
    for (final face in faces) {
      lines.add('• ${face.name} (${face.embeddingPaths.length} samples)');
    }
    return lines.join('\n');
  }

  String _buildDefaultResponse(int totalPhotos) {
    return 'I understand you\'re asking about photos. '
        'You currently have $totalPhotos photos in your library. '
        'Try asking about people, locations, scenes, or specific photos. '
        'Type "help" for a list of things I can do.';
  }

  String _buildPhotoContext(ImageAnalysis analysis) {
    final parts = <String>[];
    
    if (analysis.faces.isNotEmpty) {
      final names = analysis.faces.map((f) => f.name).where((n) => n != 'unknown').toSet();
      if (names.isNotEmpty) {
        parts.add('People: ${names.join(", ")}');
      }
    }
    
    if (analysis.scenes.isNotEmpty) {
      final sceneNames = analysis.scenes.take(3).map((s) => s.scene).toList();
      parts.add('Scene: ${sceneNames.join(", ")}');
    }
    
    if (analysis.objects.isNotEmpty) {
      final objNames = analysis.objects.take(5).map((o) => o.label).toList();
      parts.add('Objects: ${objNames.join(", ")}');
    }
    
    if (analysis.metadata?.date != null) {
      parts.add('Date: ${analysis.metadata!.date}');
    }
    
    return parts.join('; ');
  }

  String _formatDate(DateTime dt) {
    return '${_weekday(dt.weekday)}, ${_month(dt.month)} ${dt.day}, ${dt.year}';
  }

  String _formatTime(DateTime dt) {
    final hour = dt.hour > 12 ? dt.hour - 12 : (dt.hour == 0 ? 12 : dt.hour);
    final period = dt.hour >= 12 ? 'PM' : 'AM';
    return '$hour:${dt.minute.toString().padLeft(2, '0')} $period';
  }

  String _weekday(int day) {
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    return days[day - 1];
  }

  String _month(int month) {
    const months = ['January', 'February', 'March', 'April', 'May', 'June',
      'July', 'August', 'September', 'October', 'November', 'December'];
    return months[month - 1];
  }

  void clearMemory() {
    _conversationHistory.clear();
    _db.clearChatHistory();
  }

  void dispose() {
    _processor.dispose();
  }
}

File: src/chatbot.py
from pathlib import Path
from src.database import Database
from datetime import datetime
import re
import os


class ChatBot:
    def __init__(self, database: Database):
        self.db = database
        self.llm = None
        self._init_llm()

    def _init_llm(self):
        model_path = Path("models/chatbot/smollm2-135m-instruct-q4_k_m.gguf")
        if not model_path.exists():
            return
        try:
            from ctransformers import AutoModelForCausalLM
            self.llm = AutoModelForCausalLM.from_pretrained(
                str(model_path),
                model_type="llama",
                max_new_tokens=256,
                temperature=0.7,
            )
        except Exception:
            self.llm = None

    def chat(self, message: str) -> str:
        if self.llm is not None:
            return self._llm_chat(message)
        return self._rule_based_chat(message)

    def _llm_chat(self, message: str) -> str:
        try:
            context = self._build_context(message)
            prompt = (
                "You are a helpful photo assistant. You have access to the user's photo library. "
                "Answer concisely based on the library data provided.\n\n"
                f"Library data:\n{context}\n\n"
                f"User: {message}\nAssistant:"
            )
            output = self.llm(prompt)
            if isinstance(output, str):
                return output.strip()
            return str(output).strip()
        except Exception:
            return self._rule_based_chat(message)

    def _build_context(self, message: str) -> str:
        parts = []
        msg_lower = message.lower()

        total_photos = self.db.get_total_photo_count()
        parts.append(f"Total photos: {total_photos}")

        locations = self.db.get_all_locations()
        if locations:
            loc_names = [loc["name"] for loc in locations[:20]]
            parts.append(f"Locations found: {', '.join(loc_names)}")

        faces = self.db.get_all_faces()
        if faces:
            face_names = [f["name"] for f in faces if f["name"] != "unknown"]
            if face_names:
                parts.append(f"People identified: {', '.join(set(face_names))}")

        scenes = self.db.get_all_scenes()
        if scenes:
            scene_names = [s["name"] for s in scenes[:10]]
            parts.append(f"Scenes: {', '.join(scene_names)}")

        recent = self.db.search_photos(limit=5)
        if recent:
            recent_dates = [p.get("date_taken") for p in recent if p.get("date_taken")]
            if recent_dates:
                parts.append(f"Most recent photo dates: {', '.join(recent_dates)}")

        return "\n".join(parts)

    def _rule_based_chat(self, message: str) -> str:
        msg = message.lower()
        now = datetime.now()

        greeting_words = ["hello", "hi", "hey", "greetings"]
        if any(msg.startswith(w) for w in greeting_words):
            return (
                f"Hello! I'm your photo assistant. It's {now.strftime('%A, %B %d, %Y')}. "
                f"I can help you explore your photo library. Try asking about people, places, "
                f"or specific photos. Type /help to see what I can do."
            )

        if "help" in msg:
            return (
                "I can help you with:\n"
                "• Find photos by person: \"Show me photos with [name]\"\n"
                "• Find photos by place: \"Show photos from [location]\"\n"
                "• Find photos by time: \"Show photos from last month\"\n"
                "• Photo stats: \"How many photos do I have?\"\n"
                "• Organize photos: \"Organize my photos\"\n"
                "• Recent photos: \"Show recent photos\"\n"
                "• Scene info: \"What scenes are in my photos?\""
            )

        total = self.db.get_total_photo_count()

        if any(w in msg for w in ["how many", "total", "count", "stats", "statistics"]):
            locations = self.db.get_all_locations()
            faces = self.db.get_all_faces()
            scenes = self.db.get_all_scenes()
            parts = [f"You have {total} photos in your library."]
            if locations:
                parts.append(f"They span {len(locations)} different locations.")
            if faces:
                known = [f for f in faces if f["name"] != "unknown"]
                parts.append(f"I've identified {len(known)} different people.")
            if scenes:
                parts.append(f"There are {len(scenes)} different scene types.")
            return " ".join(parts)

        if "recent" in msg or "latest" in msg or "newest" in msg:
            photos = self.db.search_photos(limit=5)
            if photos:
                lines = ["Here are your most recent photos:"]
                for p in photos:
                    name = Path(p["file_path"]).name
                    date = p.get("date_taken") or "unknown date"
                    lines.append(f"• {name} ({date})")
                return "\n".join(lines)
            return "No photos found in the database yet."

        scene_query = ["scene", "type", "category", "indoor", "outdoor", "beach", "mountain"]
        if any(w in msg for w in scene_query):
            scenes = self.db.get_all_scenes()
            if scenes:
                lines = ["Here are the scene types in your photos:"]
                for s in scenes:
                    count = self.db.count_photos_by_scene(s["name"])
                    lines.append(f"• {s['name'].title()}: {count} photos")
                return "\n".join(lines)
            return "No scenes analyzed yet. Run the organizer first."

        location_query = ["location", "place", "where", "city", "country"]
        if any(w in msg for w in location_query):
            locations = self.db.get_all_locations()
            if locations:
                lines = ["Here are the locations in your photos:"]
                for loc in locations:
                    lines.append(f"• {loc['name']} ({loc['photo_count']} photos)")
                return "\n".join(lines)
            return "No locations found yet. Run the organizer first."

        person_query = ["person", "people", "face", "who", "someone", "recognize"]
        if any(w in msg for w in person_query):
            faces = self.db.get_all_faces()
            if faces:
                lines = ["Here are the people I've identified:"]
                for f in faces:
                    lines.append(f"• {f['name']} ({f['face_count']} faces)")
                return "\n".join(lines)
            return "No faces identified yet. Run the organizer first."

        if any(w in msg for w in ["today", "date", "day", "time"]):
            return f"Today is {now.strftime('%A, %B %d, %Y')} at {now.strftime('%I:%M %p')}."

        if any(w in msg for w in ["hello", "hi", "hey", "how are you"]):
            return (
                f"Hey there! It's {now.strftime('%A')}. "
                f"You have {total} photos. How can I help you explore them?"
            )

        if "sort" in msg or "organize" in msg:
            return "I can help organize photos. Use the Organize button in the dashboard to start sorting."

        if any(w in msg for w in ["thank", "thanks", "nice"]):
            return "You're welcome! Let me know if you need anything else with your photos."

        return (
            f"I understand you're asking about photos. "
            f"You currently have {total} photos in your library. "
            f"Try asking about people, locations, scenes, or specific photos. "
            f"Type /help for a list of things I can do."
        )
